From 0e814aadd06ee40a99d5fedd2314c9b16bc78db5 Mon Sep 17 00:00:00 2001
From: Jiri Benc <jbenc@redhat.com>
Date: Fri, 12 Sep 2014 14:30:23 -0400
Subject: [net] openvswitch: distinguish between the dropped and consumed skb

Message-id: <c501277f50c89e35b813728f7fa0256480b0de0f.1410531783.git.jbenc@redhat.com>
Patchwork-id: 94016
O-Subject: [RHEL7.1 net PATCH 50/50] openvswitch: distinguish between the dropped and consumed skb
Bugzilla: 1110384
RH-Acked-by: Flavio Leitner <fbl@redhat.com>

Bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1110384
Brew: https://brewweb.devel.redhat.com/taskinfo?taskID=7948291

commit c5eba0b6f84eb4f0fdc1d8a4abc1c7d40db6e8a6
Author: Li RongQing <roy.qing.li@gmail.com>
Date:   Wed Sep 3 17:43:45 2014 +0800

    openvswitch: distinguish between the dropped and consumed skb

    distinguish between the dropped and consumed skb, not assume the skb
    is consumed always

    Cc: Thomas Graf <tgraf@noironetworks.com>
    Cc: Pravin Shelar <pshelar@nicira.com>
    Signed-off-by: Li RongQing <roy.qing.li@gmail.com>
    Acked-by: Pravin B Shelar <pshelar@nicira.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/openvswitch/datapath.c b/net/openvswitch/datapath.c
index 31288fb..1e14d67 100644
--- a/net/openvswitch/datapath.c
+++ b/net/openvswitch/datapath.c
@@ -265,8 +265,11 @@ void ovs_dp_process_received_packet(struct vport *p, struct sk_buff *skb)
   upcall.key = &key;
   upcall.userdata = NULL;
   upcall.portid = ovs_vport_find_upcall_portid(p, skb);
-  ovs_dp_upcall(dp, skb, &upcall);
-  consume_skb(skb);
+  error = ovs_dp_upcall(dp, skb, &upcall);
+  if (unlikely(error))
+   kfree_skb(skb);
+  else
+   consume_skb(skb);
   stats_counter = &stats->n_missed;
   goto out;
  }
-- 
1.7.1